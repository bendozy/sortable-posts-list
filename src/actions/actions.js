import {
  MOVE_POST,
} from '../constants';

export default (direction, post) => ({
  type: MOVE_POST,
  payload: { direction, post },
});
