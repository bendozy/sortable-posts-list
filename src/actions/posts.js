import axios from 'axios';
import {
  FETCH_POSTS_SUCCESS,
  FETCH_POSTS_ERROR,
  MOVE_POST,
} from '../constants';

export const fetchPostsSuccess = (posts) => ({
  type: FETCH_POSTS_SUCCESS,
  payload: {
    posts: posts.map((post, index) => ({ ...post, index })),
    index: Array.from({ length: posts.length }, (v, i) => i),
  },
});

export const fetchPostsError = (error) => ({
  type: FETCH_POSTS_ERROR,
  error,
});

export const fetchPosts = () => (dispatch) => axios.get('https://jsonplaceholder.typicode.com/posts')
  .then((response) => dispatch(fetchPostsSuccess(response.data.splice(0, 5))))
  .catch((error) => dispatch(fetchPostsError(error)));


export const movePost = (direction, post) => ({
  type: MOVE_POST,
  payload: { direction, post },
});
