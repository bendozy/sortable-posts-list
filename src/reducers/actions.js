import {
  MOVE_POST,
  UP,
} from '../constants';

export default (state = [], action) => {
  switch (action.type) {
    case MOVE_POST: {
      const { direction, post: { index, title, id } } = action.payload;
      const nextIndex = direction === UP ? index - 1 : index + 1;

      return [
        ...state,
        {
          title,
          postId: id,
          from: index,
          to: nextIndex,
          key: `${id}-${index}-${nextIndex}`,
        },
      ];
    }
    default:
      return state;
  }
};
