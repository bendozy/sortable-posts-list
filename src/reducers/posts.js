import {
  FETCH_POSTS_SUCCESS,
  FETCH_POSTS_ERROR,
} from '../constants';

export const initialState = {
  posts: [],
  errors: null,
  isLoading: true,
};

export default (state = initialState, action) => {
  switch (action.type) {
    case FETCH_POSTS_SUCCESS:
      return {
        ...state,
        posts: action.posts,
        isLoading: false,
      };
    case FETCH_POSTS_ERROR:
      return {
        ...state,
        errors: action.errors,
        isLoading: false,
      };
    default:
      return state;
  }
};
