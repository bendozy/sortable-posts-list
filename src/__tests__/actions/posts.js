import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import axios from 'axios';
import { movePost, fetchPosts } from '../../actions/posts';
import {
  FETCH_POSTS_SUCCESS,
  FETCH_POSTS_ERROR,
  MOVE_POST,
} from '../../constants';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

jest.mock('axios');

describe('Posts action tests', () => {
  it('fetches post api data and dispatches first 5', () => {
    const response = {
      data: Array.from({ length: 50 }, (v, i) => ({ id: i })),
    };

    const store = mockStore({
      posts: {
        posts: [],
      },
    });

    axios.get.mockResolvedValue(response);

    return store.dispatch(fetchPosts()).then(() => {
      expect(store.getActions())
        .toContainEqual({ type: FETCH_POSTS_SUCCESS, posts: response.data.slice(0, 5) });
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  it('dispatces error action if there is an issue with the api call', () => {
    const error = {
      data: 'api error fetching data',
    };

    const store = mockStore({
      posts: {
        error: null,
      },
    });

    axios.get.mockRejectedValue(error);

    return store.dispatch(fetchPosts()).then(() => {
      expect(store.getActions())
        .toContainEqual({ type: FETCH_POSTS_ERROR, error });
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  it('dispatches movePost action', () => {
    const payload = {
      from: 1,
      to: 2,
    };
    const expectedAction = {
      type: MOVE_POST,
      payload,
    };

    expect(movePost(payload)).toEqual(expectedAction);
  });
});
