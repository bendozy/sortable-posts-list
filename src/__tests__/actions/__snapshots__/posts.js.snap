// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Posts action tests dispatces error action if there is an issue with the api call 1`] = `
Array [
  Object {
    "error": Object {
      "data": "api error fetching data",
    },
    "type": "FETCH_POSTS_ERROR",
  },
]
`;

exports[`Posts action tests fetches post api data and dispatches first 5 1`] = `
Array [
  Object {
    "posts": Array [
      Object {
        "id": 0,
      },
      Object {
        "id": 1,
      },
      Object {
        "id": 2,
      },
      Object {
        "id": 3,
      },
      Object {
        "id": 4,
      },
    ],
    "type": "FETCH_POSTS_SUCCESS",
  },
]
`;
